File: backend/app/experiments/vlm/frame_selection.py
"""Frame selection strategies (M4 §16-19, §20, §22).

- ``single-quality-v1``: the single quality-selected frame only.
- ``temporal-triad-v1``: 3 quality-eligible frames nearest 25% / 50% / 75% of burst chronology.

Callers pass only eligible frames; if fewer than 3 eligible frames exist, the triad returns what
is available and the caller must return an experiment precondition result instead of fabricating
frames (M4 §18).
"""

from __future__ import annotations

from typing import Literal

FrameStrategy = Literal["single-quality-v1", "temporal-triad-v1"]

FRAME_STRATEGIES: dict[str, int] = {
    "single-quality-v1": 1,
    "temporal-triad-v1": 3,
}


def is_supported_strategy(strategy: str) -> bool:
    return strategy in FRAME_STRATEGIES


def expected_frame_count(strategy: str) -> int:
    return FRAME_STRATEGIES[strategy]


def select_single_frame[T](items: list[T]) -> list[T]:
    """Deterministic middle-item selection (M4 §17)."""
    if not items:
        return []
    return [items[len(items) // 2]]


def select_temporal_triad[T](items: list[T]) -> list[T]:
    """Pick the items nearest 25/50/75% of the (chronological) eligible list (M4 §18)."""
    n = len(items)
    if n == 0:
        return []
    if n < 3:
        return list(items)
    used: set[int] = set()
    for target in (0.25, 0.5, 0.75):
        index = round(target * (n - 1))
        while index in used:
            index = min(n - 1, index + 1)
        used.add(index)
    return [items[index] for index in sorted(used)]

File: frontend/src/features/experiment/frameSelection.ts
/**
 * Frame selection for VLM upload (M4 §16-19, §22).
 *
 * Only quality-eligible frames are considered. single-quality-v1 sends the quality-selected frame;
 * temporal-triad-v1 sends the 3 eligible frames nearest 25/50/75% of burst chronology. If fewer
 * than 3 eligible frames exist, the triad returns [] and the caller reports a precondition result
 * rather than fabricating frames (M4 §18).
 */

import type { CaptureBundle, CaptureFrame } from '../capture/types/capture'
import type { BundleQualityAssessment } from '../capture/quality/types/quality'

export type ExperimentFrameStrategy = 'single-quality-v1' | 'temporal-triad-v1'

export function selectFramesForStrategy(
  bundle: CaptureBundle,
  quality: BundleQualityAssessment | null,
  strategy: ExperimentFrameStrategy,
): CaptureFrame[] {
  const eligibleIds = new Set(quality?.eligibleFrameIds ?? [])
  const eligible = bundle.frames
    .filter((frame) => eligibleIds.has(frame.id))
    .sort((a, b) => a.sequence - b.sequence)

  if (strategy === 'single-quality-v1') {
    const representative = bundle.frames.find((frame) => frame.id === bundle.representativeFrameId)
    return representative ? [representative] : eligible.slice(0, 1)
  }

  if (eligible.length < 3) return []
  const targets = [0.25, 0.5, 0.75]
  const picked: number[] = []
  for (const target of targets) {
    let index = Math.round(target * (eligible.length - 1))
    while (picked.includes(index)) {
      index = Math.min(eligible.length - 1, index + 1)
    }
    picked.push(index)
  }
  return picked.sort((a, b) => a - b).map((index) => eligible[index])
}
